src/WebJobs.Script/Scale/HostPerformanceManager.cs

```diff
@@ -27,20 +41,136 @@ public HostPerformanceManager(IEnvironment environment, IOptions<HostHealthMonit
         }
 
-        internal static bool IsUnderHighLoad(ApplicationPerformanceCounters counters, Collection<string> exceededCounters = null, float threshold = HostHealthMonitorOptions.DefaultCounterThreshold)
+        /// <summary>
+        /// Check both sandbox enforced performance counters as well as process level thresholds
+        /// like CPU, memory, etc.
+        /// </summary>
+        public virtual async Task<bool> IsUnderHighLoad(Collection<string> exceededCounters = null, ILogger logger = null)
+        {
+            return PerformanceCountersExceeded(exceededCounters, logger) ||
+                await ProcessThresholdsExceeded(exceededCounters, logger);
+        }
+
+        public async Task<IActionResult> TryHandleHealthPingAsync(HttpRequest request, ILogger logger)
+        {
+            var healthPingEnabled = _environment.GetEnvironmentVariableOrDefault(EnvironmentSettingNames.HealthPingEnabled, "1");
+            if (healthPingEnabled.Equals("0"))
+            {
+                // disabled at app level
+                return null;
+            }
+
+            bool checkHealth = false;
+            var userAgent = request.GetHeaderValueOrDefault("User-Agent");
+            if (!string.IsNullOrEmpty(userAgent) &&
+                (userAgent.IndexOf(ScriptConstants.HttpScaleUserAgent, StringComparison.OrdinalIgnoreCase) != -1 ||
+                userAgent.IndexOf(ScriptConstants.ScaleControllerUserAgent, StringComparison.OrdinalIgnoreCase) != -1))
+            {
+                // for these user agents, we default to true
+                checkHealth = true;
+            }
+
+            var query = request.GetQueryCollectionAsDictionary();
+            if (query.TryGetValue(ScriptConstants.HealthCheckQueryParam, out string value))
+            {
+                // header overrides user agent
+                checkHealth = value.Equals("1");
+            }
+
+            if (checkHealth)
+            {
+                // check host + worker health
+                int statusCode = (int)HttpStatusCode.OK;
+                if (await IsUnderHighLoad(logger: logger))
+                {
+                    statusCode = 429;
+                }
+
+                return new StatusCodeResult(statusCode);
+            }
+
+            return null;
+        }
+
+        internal async Task<bool> ProcessThresholdsExceeded(Collection<string> exceededCounters = null, ILogger logger = null)
+        {
+            var hostProcessStats = _processMonitor.GetStats();
+            if (hostProcessStats.CpuLoadHistory.Any())
+            {
+                string formattedLoadHistory = string.Join(",", hostProcessStats.CpuLoadHistory);
+                logger?.LogDebug($"[HostMonitor] Host process stats: EffectiveCores={_environment.GetEffectiveCoresCount()}, CpuLoadHistory=({formattedLoadHistory}), AvgLoad: {Math.Round(hostProcessStats.CpuLoadHistory.Average())}, MaxLoad: {Math.Round(hostProcessStats.CpuLoadHistory.Max())}");
```

Comment:
Can you write these log messages with an EventId, giving it a useful name? That increases the query-ability in kusto.

Reply:
I've added Logger.Defines for all of these with EventIds. Will be in next iteration